### ISSUE TITLE
Add Singleton support for Functions

### ISSUE THREAD

**Issue by mathewc** [Azure/azure-webjobs-sdk-script#912]
We should discuss whether we want to bring this functionality forward for Functions. However, it will be relatively simple for us to do - we just need to expose new properties via function.json for function level singletons. See [here](https://github.com/Azure/azure-webjobs-sdk/wiki/Singleton) for singleton doc.

In addition to Function level singleton support, we might also consider supporting Listener level Singleton as well.

**Comment by alrod** [comment]
Customer wants this feature:
https://social.msdn.microsoft.com/Forums/en-US/4f21e730-e4f3-4f5f-a408-56e23283fb92/run-azure-function-as-singleton-listener-with-node-js?forum=AzureFunctions

**Comment by vladkosarev** [comment]
We would love to have that. Let us specify singleton in different scopes just like in web jobs.

**Comment by mathewc** [comment]
Thanks @vladkosarev. This issue has been on the back burner, since not many people have been asking for it yet. What specifically is your scenario?

**Comment by vladkosarev** [comment]
Being able to process messages from multiple queues in series as long as they affect the same entity. So if we have two queues and both of them will affect Order/1 (which is encoded in queue messages) we want the two different functions to kick in in series and process messages from those two queues one by one instead of in parallel. The idea here is having serialized access to data storage for a particular entity.

**Comment by ajukraine** [comment]
@vladkosarev sounds like actor model. I would love to see how efficient it's with using scoped singleton and Azure functions. Another option is to use Service Fabric, but it's huge piece of infrastructure to be handled. Prefer to start with something as lightweight as Azure functions

**Comment by vladkosarev** [comment]
That is exactly what I'm trying to achieve. Actor-like model using Azure functions. Now that they announced that functions will support serverless framework this might not be as important but I'd still like this ability in the 'raw'. Service Fabric is great but it's still not serverless. I want to pay for compute resources not for VMs. Consumption functions + actors on top is my path to nirvana. Obviously you can't have in memory state, etc but it would still be a good start to the path of properly infinitely scalable architecture.

**Comment by WonderPanda** [comment]
@lindydonna Hey I'm just wondering if there are any updates in regards to when we might expect support for singleton functionality with Azure Functions. Based on the conversation here it seemed like it might be low hanging fruit. It would be extremely helpful to be able to specify singleton behavior in function.json

**Comment by tmakin** [comment]
I have use case where this would very helpful. I have long running background processing tasks which are triggered via a storage queue. In order to prevent spikes in the database load, I need to ensure that that there is not more than one queue item being processed at a time. 

My current solution is to use a timer trigger with a short interval to manually poll the queue, but a singleton flag for queue triggers would be a much tidier option.

**Comment by ericleigh007** [comment]
Yes, I didn't know what this was from the title.  Perhaps rename the report to something more descriptive.

My case is just about the same.  Would like to guarantee only one queue function running at a time.  NOTE with the current time limitations we cannot just wait.

-thanks Donna
-e

**Comment by alohaninja** [comment]
Seems like you could support locking on your own - you just need shared storage backing it - SQL Azure, Azure Blob/Table/Queue, Redis, etc. Would be great just to add a `[SingletonAttribute]` to our Azure Functions like webjobs has. [`host.json` has configurable options for singletons](https://github.com/Azure/azure-webjobs-sdk-script/wiki/host.json), but I don't know if it supports Azure functions (AF) versus Web Jobs. Could just add an environment key which has the storage connection string etc, and assign it in our `host.json`.

**Comment by aboersch** [comment]
@lindydonna any updates on a timeline for this feature?

I have a couple of Projects where I would like to switch from WebJobs to Azure Functions as well as some new Projects that need serial processing for queues, which require this functionality.

From my understanding of the way this works for WebJobs is a lock blob with a lease in the storage accounts. Azure Functions appear to already use this mechanism for the Timer-Trigger.

**Comment by aboersch** [comment]
@alohaninja Supporting locking on our own is not trivial. e.g. In a Queue-Trigger Function you could only throw an exception so that the message is put back in the queue, this may however lead to the message being marked as poison and therefore lost if you cannot process it in time. Additionaly the Function will still be invoked leading to extra costs.

According to this [issue](https://github.com/Azure/Azure-Functions/issues/422) there is currently no support for the Singleton Attribute in Azure Functions and the host.json options are therefore moot.

Possible Workarounds:
- Timer-Trigger Functions appear to run as Singletons. They produce a blob in the storage account under "locks/" and create a lease. This requires implementing the queue Mechanism on your own.
- [ServiceBus-Trigger Functions](https://docs.microsoft.com/en-us/azure/azure-functions/functions-bindings-service-bus#trigger-behavior) with serviceBus.maxConcurrentCalls to 1 in host.json. This is however a global setting and I would like use this on a per function basis.

@lindydonna is it possible to confirm that Timer-Trigger Functions run as singletons?

**Comment by alohaninja** [comment]
@aboersch - came here because we are using **Timer-Trigger** functions and they can run simultaneously, I was looking for a way to ensure you cannot have concurrent timer events - seems to occur during app restart (_DLLs change in `/bin`, app restart via portal_).

Configured `function.json` to ignore trigger firing @ restarts via  `"runOnStartup": false`, but we still see concurrent executions if a function was running before the cycle. Seems like the locking doesn't account well for function updates (restart events) - it fires the trigger even though an existing process is already running. To verify this - use kudu process explorer and you'll see multiple processes for the same function. 

For now - I just use kudu process explorer to kill any existing processes before making any app updates or restarting the function host - would be great if the portal allowed you to kill the process for a running azure function.

**Comment by aboersch** [comment]
@alohaninja I am aware of the troubles with the restart. I usually stop the app before updating because of it, however all my functions are designed to be interrupted at any time.

If you look into your storage account you will see a container called azure-webjob-hosts. There will be several folders here <hostname>-<random-number> which contain a host.Functions.<function-name>.listener file for each timer-trigger function. This file is being used to lock with a blob lease.

Every time your app is (re)started a new folder is created (<hostname>-<random-number>). Since the new folder is empty there is no blob and hence no lease to check for, hence the parallel execution.

This should perhaps be a separate issue though.

**Comment by rossdargan** [comment]
I could really do with this feature to. The issue I have is that I need to call an external api based to get more data any time a message gets added to a queue. I tend to get around 300+ messages over a few minutes every 8 hours. The issue I'm having is azure spins up 16 servers to handle the spike in messages (which is cool...) however this is utterly destroying the server I'm calling.

I've set batch size and singleton in the host.json but that appears to have no impact (setting batch size just results in more servers being started).

**Comment by BowserKingKoopa** [comment]
I'm in desperate need of this as well. I have a bunch of webjobs I'd like to move to Azure Functions, but I can't because they need to run as singletons. Some are queue based but need to be run in order. Others call external apis that are very sensitive about how often I call them.

**Comment by rossdargan** [comment]
To work around this I made a buffer queue and use a scheduled function to see how many items are in the processing queue and move a few items over depending on the count.

**Comment by WonderPanda** [comment]
@BowserKingKoopa @rossdargan I haven't had time to experiment with it yet but in the configuration settings for host.json there's an option for `WEBSITE_MAX_DYNAMIC_APPLICATION_SCALE_OUT`. Sounds like it's still being worked on but using this in conjunction with batch size might help you achieve singleton semantics for the meantime.

**Comment by bwwilliam** [comment]
Any updates on this please? Has anyone found a way to make enforce Functions run as a single instance max.

**Comment by KeenFann** [comment]
We also have a need for this. Both singleton and scale out limitation at function level.

**Comment by ajukraine** [comment]
I guess, if you can't wait for it to be implemented in SDK, you can use it already in Durable Azure Functions (https://docs.microsoft.com/en-us/azure/azure-functions/durable-functions-overview)

**Comment by AntonChernysh** [comment]
+1.
Use case: Simple slack bot event handler, should only send message once.

**Comment by WonderPanda** [comment]
@AntonChernysh I think you might be confused about Singleton behavior... There is nothing today preventing you from building a slack bot that only responds to messages only once

**Comment by AntonChernysh** [comment]
@WonderPanda looks like my function is scaling and running multiple times, therefore getting reply as many times as functions started. I have the same function  (python 3.6) running on AWS lambda.
I'd appreciate If you could advice something to make function run only once.

**Comment by WonderPanda** [comment]
@AntonChernysh What triggers your function?

**Comment by bwwilliam** [comment]
Any idea when singleton can be made available please? I'm implementing CQRS pattern with functions. My event publisher needs to be singleton so it can process the events in the right order/sequence. Thanks

**Comment by AntonChernysh** [comment]
post message to function's HTTPs endpoint. Can we continue in skype? anton.chernysh

**Comment by alexjustus** [comment]
We would love this ability too.

The Consumption plan for Functions is incredibly enticing, but since we need to consume our queue messages one at a time, we're limited to web jobs only for now.

**Comment by Schenz** [comment]
+1

I too have the need to process queue items one at a time in order, so a singleton is exactly what is needed.

**Comment by tony-gutierrez** [comment]
Apparently this functionality exists, it just hasn't been documented for anything other that .net

https://github.com/Azure/azure-webjobs-sdk/wiki/Singleton

**Comment by paulbatum** [comment]
@tony-gutierrez that functionality is for webjobs. This issue is about bringing that functionality to functions.

**Comment by tony-gutierrez** [comment]
As far as I can tell, functions issues get handled by that team, but I have no idea really. MS' documentation and organization is a mess.

**Comment by paulbatum** [comment]
I'm sorry to hear you feel that way. Yes we are the same team for functions and webjobs, but there is a clear layering difference, not all webjobs features light up in functions. Often additional work is required, as is the case here.

**Comment by rossdargan** [comment]
@paulbatum Has this been closed because it's been resolved or closed because it won't be added as a feature?

**Comment by tony-gutierrez** [comment]
I think this is supposed to be the tracking issue: https://github.com/Azure/azure-webjobs-sdk-script/issues/912

**Comment by tony-gutierrez** [comment]
It's not closed.

**Comment by Hallmanac** [comment]
Is there a place other than this issue to track this feature? This is something we are in need of as well. Just like many others here, we have a queue that we need to handle synchronously and allowing the function to spin up as many instances as it needs is problematic for that use case.

**Comment by paulbatum** [comment]
This issue is where its tracked.

**Comment by joeizy** [comment]
I could really use this right now. Until this is implemented I may need to resort to something like Event Hubs or Service Bus queues. Sad...

Created feature request: https://feedback.azure.com/forums/355860-azure-functions/suggestions/32914021-sequential-processing-of-storage-queues

**Comment by thuansoldier** [comment]
Already voted for synchronous in Queue message. Note that setting batch size to 1 does not support sequential processing. Batch size is to let Azure Function to process 1 by 1.

**Comment by SimonLuckenuik** [comment]
What about this blog post? https://medium.com/@yuka1984/azure-functions-の-singletonattributeとmodeプロパティ-bb728062198e

**Comment by SimonLuckenuik** [comment]
Some warnings on usage in this thread: https://twitter.com/cgillum/status/967158071091871746?s=21

**Comment by ThomasGuttesen** [comment]
I have had the same issue in counter like scenario.
I came up with poc like this with Durable Functions and Eternal orchestrations. 
Since DurableOrchestrationContext runs in a single thread, it seems doable to control the FIFO data from the cloud queue. 
If you sort on the Time column in the test1 table,  you see RowKey is in the same order as the queue delivers.

namespace CheckSingleton
{
    public static class DurableSingeltonTest
    {
        private static Lazy<Guid> partitionKey = new Lazy<Guid>(() => Guid.NewGuid());

        public class PocoEntity : TableEntity
        {
            public int Number { get; set; }
            public long Time { get; set; }
            public int ItemsInBatch { get; set; }
        }

        [FunctionName("Main")]
        public static async Task Main(
            [OrchestrationTrigger]  DurableOrchestrationContext ctx,
            TraceWriter log)
        {
            var t = ctx.CallSubOrchestratorAsync("QueueTrigger", "test1");
            await ctx.CallActivityAsync("Test1", "test1");
        }


        [FunctionName("QueueTrigger")]
        public static async Task QueueTrigger(
            [OrchestrationTrigger] DurableOrchestrationContext context,
                IBinder binder
            )        {

            string queueName = context.GetInput<string>();

            bool waitingPeriod = true;
            try
            {
                var collection = await context.CallActivityAsync<List<string>>("CheckQueue", queueName);
                if (collection.Count > 0) {
                    await context.CallActivityAsync("AddToTable", Tuple.Create(queueName, collection));
                    waitingPeriod = false;
                }
            }
            catch (Exception)
            {

                throw;
            }

            if (waitingPeriod)
            {
                DateTime nextCleanup = context.CurrentUtcDateTime.AddSeconds(10);
                await context.CreateTimer(nextCleanup, CancellationToken.None);
            }

            context.ContinueAsNew(queueName);
        }

        [FunctionName("CheckQueue")]
        public static async Task<List<string>> CheckQueue(
    [ActivityTrigger]  DurableActivityContext ctx,
    IBinder binder,
    TraceWriter log)
        {

            var name = ctx.GetInput<string>();

            var outputQueue = await binder.BindAsync<CloudQueue>(new QueueAttribute(name));

            QueueRequestOptions opt = new QueueRequestOptions() { MaximumExecutionTime = TimeSpan.FromSeconds(30) };

            var m = await outputQueue.GetMessagesAsync(16, TimeSpan.FromSeconds(30), opt, null);
       
            List<string> list = new List<string>();

            try
            {
                foreach (var item in m)
                {
                    await outputQueue.DeleteMessageAsync(item);
                    list.Add(item.AsString);
                }
            }
            catch (Exception)
            {
                //Just ignore;
            }


            return list;
        }


        [FunctionName("AddToTable")]
        public static async Task AddToTable(
    [ActivityTrigger]  Tuple<string, List<string>> input,
    IBinder binder,
    TraceWriter log)
        {

            var tableBinding = await binder.BindAsync<ICollector<PocoEntity>>(new TableAttribute(input.Item1));


            foreach (var item in input.Item2)
            {
                var p = new PocoEntity()
                {

                    PartitionKey = partitionKey.Value.ToString("N"),
                    RowKey = item.ToString(),
                    Time = DateTime.UtcNow.Ticks,
                    ItemsInBatch = input.Item2.Count
                };

                tableBinding.Add(p);
            }


        }

        [FunctionName("Test1")]
        public static async Task Test1(
            [ActivityTrigger]  string name,
            IBinder binder,
            TraceWriter log)
        {

            var queue = await binder.BindAsync<CloudQueue>(new QueueAttribute(name));

            for (int i = 0; i < 100; i++)
            {
                await queue.AddMessageAsync(new CloudQueueMessage(i.ToString()));
            }

        }

    }
}

**Comment by SimonLuckenuik** [comment]
@mathewc, @jeffhollan are there any plan to support Singleton with Consumption plan at all? 

Can you confirm that the only impact of [Singleton] in a consumption plan is around billing (being charged while waiting for the singleton lock to be released)? (Not sure I understand why the Blob lease approach would not work specifically in consumption plan).

I would love paying to wait if it provides me with an easy/ready to use Singleton pattern. Otherwise it means, re-implementing the Blob lease approach that was already implemented by for Webjobs, documented [here](https://github.com/Azure/azure-webjobs-sdk/wiki/Singleton) and executing it from my Function. 

This Singleton topic has been floating around for a while, maybe officially testing it and supporting it would remove some confusion and enable new scenarios. Functions is built over Webjobs, so initial assumption by developers is probably that if it works in Webjobs, it will work for Functions.

**Comment by jeffhollan** [comment]
Longer thread and want to make sure I understand and also share what may be possible today.  From basic understanding of thread it sounds like the most common scenario is something like "I want to process queue items 1 by 1 because queue message A needs to be completed before queue message B starts."  

We do allow you to control concurrency to a fair degree *per instance* with the [host.json](https://docs.microsoft.com/en-us/azure/azure-functions/functions-host-json) file.  Here you could set something like Service Bus max concurrency to 1 with 0 prefetch so each instance was only processing 1 message at a time (obviously with impacts to throughput - as per singleton as a whole).

The one piece that the host.json doesn't control is the concurrency across all possible instances.  If you do something like drop 1000 messages in a queue it's very likely we will scale out a few instances as competing consumers to help drain the queue.  There is the [WEBSITE_MAX_DYNAMIC_APPLICATION_SCALE_OUT](https://github.com/projectkudu/kudu/wiki/Configurable-settings#limit-the-scaling-of-function-apps) app setting that does let you limit scale out. It's not perfect, but for small values (like less than 5) it should work well.  We also have a work item to make this work for any value.

That said I think if you set the correct host.json value and WEBSITE_MAX_DYNAMIC_APPLICATION_SCALE_OUT=1 you would only have 1 instance processing 1 message at a time.

Even better - @SimonLuckenuik  you mentioned you aren't too worried about cost. If you don't use a consumption plan but deploy to app service plan you can fully control scale and fix your function at 1 instance reliably.  So you could always do host.json + 1 instance in an app service plan and get this too.

Not sure if a scenario here that a mix of host.json / instance restrictions won't solve for though so let me know.  Also worth noting - [storage queues don't guarantee ordering](https://docs.microsoft.com/en-us/azure/service-bus-messaging/service-bus-azure-and-service-bus-queues-compared-contrasted#foundational-capabilities) so if your scenario is impacted by ordering may want to consider something like service bus.

**Comment by spzSource** [comment]
Hi all, @jeffhollan 
Reading whole discussion didn't give me a clear understanding about does singleton functions (Singleton attribute) officially supported for Azure Function under App Service plan.

Could you clarify can we use [Singleton] attribute for Azure Function under App Service plan in production?

**Comment by jeffhollan** [comment]
I'm working to understand current behaviors fully but as called out [here](https://github.com/Azure/azure-functions-host/issues/1633#issuecomment-402311376) it appears Singleton should "just work" for pre-compiled C# functions today, and this specific issue is more about making it possible from other languages (Node, Java, .csx).  There is the unknown of if there may be billing implications (time spent waiting for lock to release may be billed).  I'll work to confirm this though and apologies for confusion (on my part as well) on supportability of this feature

**Comment by WonderPanda** [comment]
I love you Jeff

**Comment by Schaemelhout** [comment]
In the meantime, would it be a good idea to have 1 queue (**A**) for dumping a bunch of messages while having another queue (**B**) to use as the actual trigger?
I could process the messages 1 by 1 by dequeuing a message from (**A**) and queuing it on the 'trigger' queue (**B**) when a message has been processed.

**Comment by christian-vorhemus** [comment]
@jeffhollan, I tried to follow your suggestions, however I experience some difficulties. Here is the scenario I'm working on: I'm quickly loading thousands of images into a Blob (A). An Azure Function (AF) gets triggered by a BlobTrigger and adds the Blob image path enriched with some additional information into a ServiceBus Queue (C). Another AF (D) started by a ServiceBusTrigger fetches the messages from the Queue and calls a Third-Party Web API (E) which is very slow. Finally, the results of (E) are stored in (F)

![architecture](https://user-images.githubusercontent.com/3229424/44288840-06de3f00-a272-11e8-932f-c32f985ea59c.png)

It is necessary to throttle AF (D) to not overload (E). However, (D) gets flooded by messages from the Queue Trigger, even though I set maxConcurrentCalls to 1 in host.json and WEBSITE_MAX_DYNAMIC_APPLICATION_SCALE_OUT to 1. (D) approximately takes 2 minutes to finish since the call to (E) takes a while.

Maybe my architecture is wrong and there is a clever and easy way to do this?

**Comment by simeyla** [comment]
Ugh. So why is it called a `QUEUE` then.

Should be called a `MassiveDumpOfStuffThatNeedsDoingInNoParticularOrder`

It's always the simplest of concepts that become the most complex - every time. 
I should know this by now.

**Comment by simeyla** [comment]
@jeffhollan do we know if Singleton guarantees FIFO order? I suspect that effectively it would (certainly for my needs it would), but I'm not sure if it is guaranteed.

**Comment by spzSource** [comment]
@simeyla There is no guarantees for FIFO order even using Singleton attribute. The only way is to set queue.batchSize to 1 along with Singleton attribute for achieving your needs.

**Comment by simeyla** [comment]
###PSA: Singleton does not work for consumption plan. It just fails silently.

Yes, others in this thread have mentioned it but it's easy to miss.

**Comment by jeffhollan** [comment]
Few things as I just tested this out.  first thing I noticed:

1. `[Singleton]` does not appear to work when scaled out.  I dropped 1000 items in a queue and after it scaled out to a few instances some instances would be processed at the same time.
1. It should be noted that [Azure Storage Queues don't guarantee ordering](https://docs.microsoft.com/en-us/azure/service-bus-messaging/service-bus-azure-and-service-bus-queues-compared-contrasted#foundational-capabilities) regardless of tech you have consuming.

I believe in the interm the only way I can think of this that would work is setting batch size to 1 and setting WEBSITE_MAX_DYNAMIC_APPLICATION_SCALE_OUT to 1.  Possible setting batch size to 1 and singleton have same impact, however I don't know if singleton would introduce higher billing as singleton does dequeue a bunch of messages, just holds them for a long time - which holding time I can't confirm if billed or not.

@christian-vorhemus I see you said the above (WEBSITE_MAX_DYNAMIC_APPLICATION_SCALE_OUT + batch size) *wasn't* working but not sure how. The queue getting flooded sounds expected, the main thing is that D is taking ages because you are only processing one item at a time.

As an aside 2 features we are looking at that should help with this longer term:

1. A more reliable way to limit scale out to just 1 instance (WEBSITE_MAX_DYNAMIC_APPLICATION_SCALE_OUT has some quirks, especially if set to a value > 5)
2. Look into if we can support service bus Sessions. This could enable faster processing across multiple instances while still preserving order.

**Comment by spzSource** [comment]
> [Singleton] does not appear to work when scaled out. I dropped 1000 items in a queue and after it scaled out to a few instances some instances would be processed at the same time.

Have you tried [Singleton(Model = SingletonMode.Listener)]? Documentation says that using this mode, the listener will only be running on a single instance (when scalled out).

So, as far as I understood, in case  scalled out env function definition must be the following:
```cs
[Singleton(Mode = SingletonMode.Function)]
[Singleton(Mode = SingletonMode.Listener)]
public static void Run()
{
}
```

**Comment by SimonLuckenuik** [comment]
Thanks @jeffhollan these are good news!

**Comment by spzSource** [comment]
@jeffhollan thanks for exhaustive answer. Can we assume that `Singleton` functionality will not be removed in future releases?

**Comment by SimonLuckenuik** [comment]
@jeffhollan I am not sure to understand why the recommendation is not to use it for consumption plan, if it is explicitly said in the documentation that you pay as soon the wait starts. I understand that you might have to keep things in mind for your scenario to see if the billing model fits you (like your 16 messages example), but paying for wait time is still cheaper compared to dedicated plan in a lot of scenarios. For example, we have low concurrency HTTP trigger scenarios where a singleton lock is a great protection to prevent overlapping operations where transactions are not available.

**Comment by jeffhollan** [comment]
@SimonLuckenuik I suppose you can use it and just pay extra - and I'd have to chat with the team but I don't imagine we'd remove it or block it. Just keep it relatively as-is and document it's behavior

**Comment by zurcacielos** [comment]
@jeffhollan Singleton attribute is not working for me in azure functions under application service plan. When the function is tarted with a trigger like this for example:
`        [FunctionName("YammerETL"), Singleton("YammerETLLock", SingletonScope.Host)]
        public static async void Run([TimerTrigger("0 */1 * * * *", RunOnStartup = true)]TimerInfo myTimer, TraceWriter log, Microsoft.Azure.WebJobs.ExecutionContext context)
{ 
// do something
}`

I tried runonstartup false, and other triggers and Singleton at the function level. But if the singletonattribute would work, even with the timertrigger I should not have two instances at the same time running. This happened first at debug time under visual studio code, then it happened when deployed to azure.

**Comment by tbdowns** [comment]
I'm trying to do something a little simpler. If the function is already being executed, I want to forcibly return HTTP 429 instead of the the request being queued. I have configured as follows but no luck...

[Singleton]

  "http": {
    "routePrefix": "api",
    "maxOutstandingRequests": 0,
    "maxConcurrentRequests": 1,
    "dynamicThrottlesEnabled": true
  }

WEBSITE_MAX_DYNAMIC_APPLICATION_SCALE_OUT = 1

**Comment by jeffhollan** [comment]
@tbdowns strange as this one ideally could be solved without the `[Singleton]` just with the host configuration you have here. May be worth opening a separate issue on why you aren't getting a 429 with those settings (only thought is potentially disabling dynamic throttles).

**Comment by CoenraadS** [comment]
Would this also work for CronJobs, to prevent another one starting is the previous one runs longer then expected?

**Comment by paulbatum** [comment]
@CoenraadS timer triggers already work this way. You should never see two executions of the same timer trigger happening concurrently - if you're seeing this its either a configuration issue or a bug and I would suggest you file a separate issue about it.

**Comment by jeffhollan** [comment]
@paulbatum I suspect the ask may be if you set a cron expression to fire every 10 seconds, but never want more than one at one time.  If execution A fires at time 0:00 and doesn't complete until 0:11.  At 0:10 execution B may "trigger" but you may not want it to run unless no other executions in flight (in this case A is in flight).  The behavior gets tricky from here because do you want B to "queue up" and start whenever A finishes (potentially just piling on executions if it always takes 10+ seconds), or just skip B and wait for 0:20.

Had some similar requests for Logic Apps back in the day and we went with the latter (skip the execution and wait for next interval), but curious on what, if any, options exist for functions

**Comment by paulbatum** [comment]
> I suspect the ask may be if you set a cron expression to fire every 10 seconds, but never want more than one at one time. If execution A fires at time 0:00 and doesn't complete until 0:11. At 0:10 execution B may "trigger" but you may not want it to run unless no other executions in flight (in this case A is in flight).

Sorry I must be missing something, but that exact ask is what I spoke to above. This is how the timer trigger works today.

> The behavior gets tricky from here because do you want B to "queue up" and start whenever A finishes (potentially just piling on executions if it always takes 10+ seconds), or just skip B and wait for 0:20.

WebJobs/Functions use the queueing approach rather than the skipping approach, but it will only queue once. So if you had a 10 second interval and you had an execution run for 35 seconds, the system would start another execution immediately, but if that execution only took a second, then the system would go idle until the next normally scheduled execution. As far as I know this behavior is not configurable.

**Comment by adrianknight89** [comment]
Guys, any update on this? For my use case, I need a way to serialize changes to globally unique fields like usernames, and I'm using CosmosDB which doesn't have uniqueness guarantees across partitions. So, the only way to achieve this, afak, is through a singleton function. This feature request should not be put on the backburner.

**Comment by SimonLuckenuik** [comment]
The singleton is working: https://github.com/Azure/azure-functions-host/issues/912#issuecomment-419608830

**Comment by fabiocav** [comment]
@jeffhollan this issue has a lot of activity and evidence that we need to enhance the feature and/or documentation for a lot of the scenarios described above. Can we (perhaps after a sync) define concrete items so this issue is actionable? If we don't believe there's any action to be taken on this (this issue suggests there is), we should close this instead with a clear description and guidance.

**Comment by Banchio** [comment]
I did some tests using functions v2 in linux container. I used a TimerTrigger function for brevity (since it is expected to be a singleton). If I start two docker containers (pointing to the same storage account) they will get two different host id so the timer function gets executed many times. Is this behaviour going to change to permit a singleton across several machines? TIA

**Comment by paulbatum** [comment]
@Banchio If you deploy this to the Azure Functions service on Linux, with an app service plan with two instances, you should see that both running containers get the same Host ID and you get singleton behavior. Similarly if you deploy to the Azure Functions Linux consumption plan and drive enough load to cause the app to scale out to multiple containers, you should still see the correct singleton behavior. 

If you're talking about scenarios outside of the Azure Functions service (such as running two containers locally on your machine), then it is expected that you explicitly configure the host IDs if you want those two containers to be considered as part of the same app and share singleton behavior. The reason for this is that we allow multiple different function apps to share the same storage account, so we can't assume that two random function apps using the same storage account should share singelton behavior - we need the host ID to tell us this.

**Comment by paulbatum** [comment]
I forgot to mention, the way you explicitly set a hostid in functions V2 is to set an environment variable:
`AzureFunctionsWebHost__hostid = abc123`

**Comment by Banchio** [comment]
Thanks, target scenario of my customer is running in aks, this is the reason why I tried on my machine first. Given that a singleton function can be deployed as part of a replicaset of 1 instance) so to obtain the same result) I was experimenting as mentioned above. I also tried to set the hostId explicitly but that seems to work only on windows host and not Linux (this is why I ended up with two different host id's). Thanks

Sent from Nine<http://www.9folders.com/>
________________________________
Da: Paul Batum <notifications@github.com>
Inviato: Giovedì 5 Settembre 2019 19:42
A: Azure/azure-functions-host
Cc: Simone Banchieri; Mention
Oggetto: Re: [Azure/azure-functions-host] Add Singleton support for Functions to ensure only one function running at a time (#912)


I forgot to mention, the way you explicitly set a hostid in functions V2 is to set an environment variable:
AzureFunctionsWebHost__hostid = abc123

**Comment by Schaemelhout** [comment]
I have 1 function app (on the consumption plan) with 3 queue-triggered functions.
I need a way to make sure only 1 queue message is processed at a time (from each queue, so the separate functions can run in parallel). I don't care about the order in which the messages are processed.

Is this possible? I see a lot of different options here like `[Singleton]`, `WEBSITE_MAX_DYNAMIC_APPLICATION_SCALE_OUT`, `batchSize` but since this issue is still open, I guess there is no official guidance on this?

I am using JavaScript but I am willing to switch to C# if it is easier to support this behavior.

**Comment by Tron1978** [comment]
Dropping by in 2020 to see if this has been worked on yet...

**Comment by joaoantunes** [comment]
Still no news about this subject?

**Comment by withinboredom** [comment]
https://github.com/Azure/azure-functions-host/issues/912#issuecomment-419608830 -- goes in-depth, not sure what you're looking for

**Comment by Tron1978** [comment]
WEBSITE_MAX_DYNAMIC_APPLICATION_SCALE_OUT was what I used to limit parallelism.

**Comment by NgSekLong** [comment]
Would like to add to this discussion, I saw this document today:
https://docs.microsoft.com/en-us/azure/logic-apps/send-related-messages-sequential-convoy

Which seems to suggest that, we can use Azure Function with Service Bus Session trigger to achieve sequential processing of message.

I have setup the following:

- Node 12
- Azure Function Premium
- Service Bus Queue, session enabled, other default
- No `WEBSITE_MAX_DYNAMIC_APPLICATION_SCALE_OUT`
- Other setting default

And a basic Function that will take a message out, wait 10 seconds, and end.

**function.json**
```
{
  "bindings": [
    {
      "queueName": "my-queue",
      "connection": "QueueRetriver",
      "name": "message",
      "type": "serviceBusTrigger",
      "direction": "in",
      "isSessionsEnabled": true
    }
  ]
}
```

 **index.js**
```
const helpers = require("../common/helpers");

module.exports = async function (context, message) {
    // Session ID: context.bindingData.messageSession.sessionId
    const sessionId = context.bindingData.messageSession.sessionId;

    await helpers.sleep(10000); // this is just to wait 10 seconds

    return;
};

/* 
module.exports.sleep = (milliseconds) => {
    return new Promise(resolve => setTimeout(resolve, milliseconds));
}
*/
```
From my testing result, messages with the same Session ID are processed one by one sequentially!

Any idea if there are any drawback to my design that I do not aware, cause this seems like a very viable method to process sequential execution. Cheers!

**Comment by paulbatum** [comment]
Yes I believe this is a viable approach. Messages sharing a session ID are processed in order, sequentially.

One thing I'll note - I am not sure whether this interacts correctly with the FUNCTIONS_WORKER_PROCESS_COUNT setting. So to be safe, I would make sure that setting is absent (or set to 1, the default).
https://docs.microsoft.com/en-us/azure/azure-functions/functions-app-settings#functions_worker_process_count

**Comment by rohitashwachaks** [comment]
Guys...
Any update on this?
I tried adding the singleton Tag on my Functions but nope, orchestration duplicates still being created!
This is what my code looks like

`
string orchestrationIdentifier = Guid.NewGuid().ToString();
                    response += $"Orchestration Identifier: {orchestrationIdentifier}\n";

                    DurableOrchestrationStatus existingInstance = await starter.GetStatusAsync(instanceId);
                    if (existingInstance == null)
                    {
                        await starter.StartNewAsync("QueueTriggerOrchestrator", instanceId, orchestrationIdentifier);
                        response += $"Started New Instance";
                    }
                    else if (existingInstance.RuntimeStatus == OrchestrationRuntimeStatus.Completed ||
                            existingInstance.RuntimeStatus == OrchestrationRuntimeStatus.Failed ||
                            existingInstance.RuntimeStatus == OrchestrationRuntimeStatus.Canceled ||
                            existingInstance.RuntimeStatus == OrchestrationRuntimeStatus.Terminated)
                    {
                        //await starter.PurgeInstanceHistoryAsync(instanceId);
                        await starter.StartNewAsync("QueueTriggerOrchestrator", instanceId, orchestrationIdentifier);
                        response += $"Restarted Earlier Instance. Ended with RunTimeStatus: {existingInstance.RuntimeStatus}";
                    }
                    else
                    {
                        response = $"FeedId: {obj.FeedId} Status: {obj.Status}.\nInstance of OrchestrationClient with ID '{instanceId}' already running. Cannot Process request\n";
                    }
`

The instanceId is a constant in this case and this snippet is in a QueueTrigger Function.

The problem is The QueueTrigger (Despite batch size being set to 1) scales up and starts consuming multiple messages simultaneously.
Now, both these instances reach the GetStatusAsync at the same time and (say the existingInstance == null ) try to start the orchestrator. 

This is NUTS!
How can you have the 2 instances with the same Instance ID! Shouldn't there be a clash or something in the InstanceHistory table, causing one to fail? But nothing of that sorts is happening when the two instances are exactly simultaneous.

I added the `[Singleton]` tag to both the QueueTriggered Function and the Orchestrator. But it doesn't work.

**Comment by stap123** [comment]
@RohitashwaChaks Does limiting the scale out not work for you, you can do it in the Azure Portal to prevent the function app scaling to more than 1 instance.

https://docs.microsoft.com/en-us/azure/azure-functions/functions-scale#limit-scale-out

**Comment by rohitashwachaks** [comment]
@stap123 I have several other functions in the same Functions App. Setting the `WEBSITE_MAX_DYNAMIC_APPLICATION_SCALE_OUT` to 1 will limit all the constituent functions.
I do not wish to do that (scalability was the primary reason why we shifted to Azure). Is there a way to restrict JUST THAT ONE function from scaling up?

**Comment by stap123** [comment]
@RohitashwaChaks Yeah we wanted the same thing. We just created a second function app and put the functions that were limited to 1 instance in that app and the other functions with default scaling in another app.

**Comment by UmairB** [comment]
@RohitashwaChaks accordiing the docs: https://docs.microsoft.com/en-us/azure/azure-functions/durable/durable-functions-singletons?tabs=csharp that should not be happening? Specifically the note:

> There is a potential race condition in this sample. If two instances of HttpStartSingle execute concurrently, both function calls will report success, but only one orchestration instance will actually start. Depending on your requirements, this may have undesirable side effects. For this reason, it is important to ensure that no two requests can execute this trigger function concurrently.

**Comment by ragzzy** [comment]
checking in on a long open issue/discussion... was there a solution for this?

**Comment by dev-bre** [comment]
is there any update for this? 
consumption plan is great, but if the function executes code which is meant to run from only one source, that plan is unmanageable and forces to move to AppService plan which way more expensive.,

**Comment by josejohny3** [comment]
is there any solution for the singleton issue?

**Comment by josejohny3** [comment]
> Guys... Any update on this? I tried adding the singleton Tag on my Functions but nope, orchestration duplicates still being created! This is what my code looks like
> 
> ` string orchestrationIdentifier = Guid.NewGuid().ToString(); response += $"Orchestration Identifier: {orchestrationIdentifier}\n";
> 
> ```
>                 DurableOrchestrationStatus existingInstance = await starter.GetStatusAsync(instanceId);
>                 if (existingInstance == null)
>                 {
>                     await starter.StartNewAsync("QueueTriggerOrchestrator", instanceId, orchestrationIdentifier);
>                     response += $"Started New Instance";
>                 }
>                 else if (existingInstance.RuntimeStatus == OrchestrationRuntimeStatus.Completed ||
>                         existingInstance.RuntimeStatus == OrchestrationRuntimeStatus.Failed ||
>                         existingInstance.RuntimeStatus == OrchestrationRuntimeStatus.Canceled ||
>                         existingInstance.RuntimeStatus == OrchestrationRuntimeStatus.Terminated)
>                 {
>                     //await starter.PurgeInstanceHistoryAsync(instanceId);
>                     await starter.StartNewAsync("QueueTriggerOrchestrator", instanceId, orchestrationIdentifier);
>                     response += $"Restarted Earlier Instance. Ended with RunTimeStatus: {existingInstance.RuntimeStatus}";
>                 }
>                 else
>                 {
>                     response = $"FeedId: {obj.FeedId} Status: {obj.Status}.\nInstance of OrchestrationClient with ID '{instanceId}' already running. Cannot Process request\n";
>                 }
> ```
> 
> `
> 
> The instanceId is a constant in this case and this snippet is in a QueueTrigger Function.
> 
> The problem is The QueueTrigger (Despite batch size being set to 1) scales up and starts consuming multiple messages simultaneously. Now, both these instances reach the GetStatusAsync at the same time and (say the existingInstance == null ) try to start the orchestrator.
> 
> This is NUTS! How can you have the 2 instances with the same Instance ID! Shouldn't there be a clash or something in the InstanceHistory table, causing one to fail? But nothing of that sorts is happening when the two instances are exactly simultaneous.
> 
> I added the `[Singleton]` tag to both the QueueTriggered Function and the Orchestrator. But it doesn't work.

is there any solution for the singleton issue?

**Comment by josejohny3** [comment]
is there any solution for the singleton issue @[rohitashwachaks](https://github.com/rohitashwachaks)?

**Comment by rohitashwachaks** [comment]
Hi everyone.
Sorry for being MIA.
I joined a masters course and left .NET development.

I remember Microsoft fixed this race condition in their latest package version.
So, I just upgraded my version and it worked.

Bit of a bummer. but I don't remember much else

**Comment by jaltin** [comment]
Hi @paulbatum, @fabiocav, @jeffhollan, @mathewc, @lindydonna and anyone else at MS that has engaged in this discussion throughout the year.

I am sorry to include you all but this discussion is over 5 years in the making. Hopefully one of you are the right person to respond?

It is such a pity that it is not possible to handle this easily in Azure functions (especially on consumption plan). Is there any chance that this will ever get implemented properly? 

### What I feel is needed
There is A LOT of comments in this thread and it has kind of gone in different directions and spread of focus. Therefore I will repeat my need  to clarify what I think will be a GREAT improvement (also the need expressed by a lot of other users in this issue):

- Running a function app on consumption plan
- Functions are triggered by Azure storage queue input bindings
- Different functions in the app has different scalability/parallel execution needs. Here are hypothetical examples:

  - Function_A
    - An image resizing func that saves a resized image in a new blob
    - Has no real limit on execution/load
    	- can be executing multiple times in parallel on one instance
    	- can scale out to as many more instances as needed (relying on Azure auto instance scale-out)
    	
  - Function_B
    - Calling an external API to fetch some data, then storing that in a blob
	- The external API can't handle more than a certain load (for this example let's say 3)
		- can be executing a maximum of 3 in parallel (or else external API will choke)
		- can scale out to multiple instances, but it is not resource heavy so can also run all 3 executions on the same instance (= Azure infrastructure can decide what i most suitable)
		
  - Function_C
    - Calling an external API to fetch some data, then doing some heavy processing of that data
	- The external API can't handle more than a certain load (for this example let's say 2)
		- can be executing a maximum of 2 in parallel (or else external API will choke)
		- is resource heavy so can not run more than one execution at a time per instance		
		
  - Function_D
	- Doing an operation that MUST be called as a singleton, e.g. updating a account balance in a database record
		- Must ensure only one execution happens at a time


There are options to control some of these things in host.json, but that will affect all functions in the app, so is too blunt of a tool.

Also, the suggestion to have separate function apps to control execution has been mentioned, but that becomes very hard if you have more functions, with different config needs (code reuse/duplication, increased deployment aspect of creating many func apps etc.). Therefore not a good alternative in my mind.

An example of how this could be handled on a function level could be in function.json

### Example for Function_A case
``` 
{
	// Other function.json values...

	"maxParallExecutionsPerInstance": 0, // 0 = no set limit
    "maxParallelExecutionsAcrossInstances": 0,
    "maxInstances": 0
}
```


### Example for Function_B case
``` 
{
	// Other function.json values...
	
	"maxParallExecutionsPerInstance": 0,
    "maxParallelExecutionsAcrossInstances": 3,
    "maxInstances": 0
}
```

### Example for Function_C case
``` 
{
	// Other function.json values...
	
	"maxParallExecutionsPerInstance": 1,
    "maxParallelExecutionsAcrossInstances": 2
}
```

### Example for Function_D case
``` 
{
	// Other function.json values...
	
    "singleton":true,
}

```

These are just conceptual ideas but you get the idea. 

Thanks!

**Comment by FP-IBM** [comment]
Still no news about this subject?

**Comment by bartlannoeye** [comment]
@jaltin I hope this get picked up some day indeed. I've been bugging support for over a month now to achieve your Function_C case for .NET (on event grid binding). Every other language has FUNCTIONS_WORKER_PROCESS_COUNT to work with.

**Comment by jaltin** [comment]
Hi all wonderful people at Microsoft and other contributors that works on this repository (@paulbatum, @fabiocav, @mathewc, @NickCraver, @kaibocai, @heppersonmicrosoft, @pragnagopa, @siddharth-ms, @liliankasem, @alrod, @balag0  among others)!

 Im posting on this issue one more time hoping that someone of you all can have a look at my previous comment https://github.com/Azure/azure-functions-host/issues/912#issuecomment-1059120967 and give a response on if you are considering this (it was after all created by one of your colleagues @mathewc).

This is becoming more and more of a PITA for us to manage in scenarios where we have functions needing strict singleton controls, and at the moment it forces us to create separate function projects all over the place.

Many of the posters here would loooooove to see something done to make our life's easier, so really hoping you can consider looking into it and try to address it in line with my suggested ideas.

Thanks!

**Comment by andrei-tofan** [comment]
It will be great if something like this will be added, I don't think is normal to do workarounds for a functionality that should be there.

**Comment by mr-davidc** [comment]
I too would love to have this kind of functionality to ensure Timer functions in particular only ever have a single instance running at a time...

**Comment by drdamour** [comment]
i find it very problematic that the docs https://learn.microsoft.com/en-us/azure/azure-functions/functions-host-json#singleton link to this issue to explain singleton behaviour...and this issue is a 100 comments none which do a great job of explaining any of this very well.

the best i found regarding singleton is https://learn.microsoft.com/en-us/azure/app-service/webjobs-sdk-how-to#singleton-attribute but this is NOT in the context of functions just the underlying webjobs...which i know are related. 

the most useful comment in this thread is https://github.com/Azure/azure-functions-host/issues/912#issuecomment-419608830 which is burried in github UX.  

the relevant info needs to make it to the docs as it's own page mostly likely

**Comment by MelGrubb** [comment]
We have several queued functions, and most of them should be running in parallel as much as possible. There is one function in particular that can't, though. It communicates with an outside service that handles concurrency... poorly. I need just that one function to be a singleton, and I thought that the Singleton attribute was going to save me. Then I read [here](https://github.com/Azure/Azure-Functions/issues/422#issuecomment-319836663) that it doesn't work on functions. But this [comment ](https://github.com/Azure/azure-functions-host/issues/912#issuecomment-419608830) says they DO work, albeit with billing implications. 

That's not what I'm seeing in my testing, though. I queued up multiple messages to my problem function and my logs show four of them starting before the first one finishes. It's a queue-triggered function, and I've tried the Singleton attribute with and without specifying a scope. It seems to just rip through the queue no matter what. 

I can't just change the host.json to a batch size of one because that will kill the performance of all the other functions that are perfectly happy running in parallel. What can I do?

**Comment by drdamour** [comment]
Fyi the way i’ve achieved singleton success is leveraging durable entities.  My triggers queue the work on SignalEntityAsync.  Entities will process one signal at a time.

if you need http triggered singleton with a response  u have to get tricky and loop GetEntityState() the entity for work completion (and when queueing probably throw in an id) and if its gonna take a while (10mins) respond with a 301 to a diff http trigger that keeps looking for the work to be done or loops again.

its def no [Singleton] but its been very successful.

**Comment by bakes82** [comment]
This doesn't seem to work for "ServiceBus" triggers, it's not auto extending the service bus lock.  While it is running in singleton mode, it's pointless if the lock isn't being extended on the ones waiting.

```c#
using System.Threading.Tasks;
using Microsoft.Azure.WebJobs;
using Microsoft.Extensions.Logging;

namespace TestFunctions;

public static class ServiceBusQueueTrigger1
{
    [FunctionName("ServiceBusQueueTrigger1")]
    [Singleton(Mode = SingletonMode.Function)]
    public static async Task RunAsync([ServiceBusTrigger("testqueue", Connection = "ServiceBusConnection")] string myQueueItem, ILogger log)
    {
        log.LogInformation($"C# ServiceBus queue trigger function processed message: {myQueueItem}");
        await Task.Delay(30000);
        log.LogInformation($"C# waited 30sec processed message: {myQueueItem}");
    }
}
```

```

[2023-09-13T21:02:13.168Z] Executing 'ServiceBusQueueTrigger1' (Reason='(null)', Id=58f691b8-d49d-49ec-887a-22eb940eb7ba)
[2023-09-13T21:02:13.168Z] Executing 'ServiceBusQueueTrigger1' (Reason='(null)', Id=0afba31e-6f8b-4a48-8c11-5e5d095d810d)
[2023-09-13T21:02:13.171Z] Trigger Details: MessageId: 83a96849befa409a82a81718d8ac0656, SequenceNumber: 12, DeliveryCount: 1, EnqueuedTimeUtc: 2023-09-13T21:03:08.6450000+00:00, LockedUntilUtc: 2023-09-13T21:04:12.5360000+00:00, SessionId: (null)
[2023-09-13T21:02:13.171Z] Trigger Details: MessageId: 5bc5995342fe40ca8714568729165ae6, SequenceNumber: 11, DeliveryCount: 1, EnqueuedTimeUtc: 2023-09-13T21:03:03.7220000+00:00, LockedUntilUtc: 2023-09-13T21:04:12.5360000+00:00, SessionId: (null)
[2023-09-13T21:02:13.172Z] Trigger Details: MessageId: 42e4ba60d32844f6b18d5699442ac0d0, SequenceNumber: 9, DeliveryCount: 1, EnqueuedTimeUtc: 2023-09-13T21:02:54.8630000+00:00, LockedUntilUtc: 2023-09-13T21:04:12.5360000+00:00, SessionId: (null)
[2023-09-13T21:02:13.174Z] Trigger Details: MessageId: 42f0a372ca164624bdb714c0502e89f5, SequenceNumber: 10, DeliveryCount: 1, EnqueuedTimeUtc: 2023-09-13T21:02:59.0660000+00:00, LockedUntilUtc: 2023-09-13T21:04:12.5360000+00:00, SessionId: (null)
[2023-09-13T21:02:13.259Z] C# ServiceBus queue trigger function processed message: 2
[2023-09-13T21:02:43.271Z] C# waited 30sec processed message: 2
[2023-09-13T21:02:43.326Z] C# ServiceBus queue trigger function processed message: 4
[2023-09-13T21:02:43.331Z] Executed 'ServiceBusQueueTrigger1' (Succeeded, Id=0afba31e-6f8b-4a48-8c11-5e5d095d810d, Duration=30220ms)
[2023-09-13T21:02:56.207Z] Executing 'ServiceBusQueueTrigger1' (Reason='(null)', Id=c012fa20-fe1d-476f-b8a1-2284ec920537)
[2023-09-13T21:02:56.209Z] Trigger Details: MessageId: 083163eaf0d64b7591f906569c87f1d2, SequenceNumber: 13, DeliveryCount: 1, EnqueuedTimeUtc: 2023-09-13T21:03:55.7240000+00:00, LockedUntilUtc: 2023-09-13T21:04:55.7400000+00:00, SessionId: (null)
[2023-09-13T21:02:59.727Z] Executing 'ServiceBusQueueTrigger1' (Reason='(null)', Id=cccf9108-34e2-4634-9382-752220b598e1)
[2023-09-13T21:02:59.729Z] Trigger Details: MessageId: 5a4b38d22a294826a5cb99a31d7213cd, SequenceNumber: 14, DeliveryCount: 1, EnqueuedTimeUtc: 2023-09-13T21:03:59.2570000+00:00, LockedUntilUtc: 2023-09-13T21:04:59.2570000+00:00, SessionId: (null)
[2023-09-13T21:03:02.479Z] Executing 'ServiceBusQueueTrigger1' (Reason='(null)', Id=3848f97d-6784-490c-b848-12cd1583c457)
[2023-09-13T21:03:02.486Z] Trigger Details: MessageId: b59d47a878ff4ca1a49ac89a8d85fa1d, SequenceNumber: 15, DeliveryCount: 1, EnqueuedTimeUtc: 2023-09-13T21:04:01.9920000+00:00, LockedUntilUtc: 2023-09-13T21:05:02.0070000+00:00, SessionId: (null)
[2023-09-13T21:03:06.040Z] Executing 'ServiceBusQueueTrigger1' (Reason='(null)', Id=d99f977c-582c-439b-a0ae-2d649ca4e7fc)
[2023-09-13T21:03:06.048Z] Trigger Details: MessageId: 1fc2e570decf4150b0f01a215645bea0, SequenceNumber: 16, DeliveryCount: 1, EnqueuedTimeUtc: 2023-09-13T21:04:05.5540000+00:00, LockedUntilUtc: 2023-09-13T21:05:05.5700000+00:00, SessionId: (null)
[2023-09-13T21:03:08.590Z] Executing 'ServiceBusQueueTrigger1' (Reason='(null)', Id=bce075e0-8e49-42b6-b271-28d40cd9fe8d)
[2023-09-13T21:03:08.592Z] Trigger Details: MessageId: b86abfa1f9b64a6682e5509c1cc1f60e, SequenceNumber: 17, DeliveryCount: 1, EnqueuedTimeUtc: 2023-09-13T21:04:08.1010000+00:00, LockedUntilUtc: 2023-09-13T21:05:08.1170000+00:00, SessionId: (null)
[2023-09-13T21:03:12.926Z] Executing 'ServiceBusQueueTrigger1' (Reason='(null)', Id=e434b194-9322-4625-b7d4-29a978625268)
[2023-09-13T21:03:12.927Z] Executing 'ServiceBusQueueTrigger1' (Reason='(null)', Id=840b2cf2-ace7-42c4-9c43-4fa06a240955)
[2023-09-13T21:03:12.931Z] Trigger Details: MessageId: 42e4ba60d32844f6b18d5699442ac0d0, SequenceNumber: 9, DeliveryCount: 2, EnqueuedTimeUtc: 2023-09-13T21:02:54.8630000+00:00, LockedUntilUtc: 2023-09-13T21:05:12.4450000+00:00, SessionId: (null)
[2023-09-13T21:03:12.932Z] Trigger Details: MessageId: 83a96849befa409a82a81718d8ac0656, SequenceNumber: 12, DeliveryCount: 2, EnqueuedTimeUtc: 2023-09-13T21:03:08.6450000+00:00, LockedUntilUtc: 2023-09-13T21:05:12.4610000+00:00, SessionId: (null)
[2023-09-13T21:03:12.926Z] Executing 'ServiceBusQueueTrigger1' (Reason='(null)', Id=eb0d75a0-a57d-4ebc-8f20-e1a19dfcc62c)
[2023-09-13T21:03:12.939Z] Trigger Details: MessageId: 5bc5995342fe40ca8714568729165ae6, SequenceNumber: 11, DeliveryCount: 2, EnqueuedTimeUtc: 2023-09-13T21:03:03.7220000+00:00, LockedUntilUtc: 2023-09-13T21:05:12.4450000+00:00, SessionId: (null)
[2023-09-13T21:03:13.362Z] C# waited 30sec processed message: 4
[2023-09-13T21:03:13.368Z] Executed 'ServiceBusQueueTrigger1' (Succeeded, Id=ca9ee9b5-77db-424e-99ee-28d14bf822c2, Duration=60269ms)
[2023-09-13T21:03:13.466Z] C# ServiceBus queue trigger function processed message: 1
[2023-09-13T21:03:13.605Z] Message processing error (Action=Complete, EntityPath=testqueue, Endpoint=MMCSharedServiceBus.servicebus.windows.net)
[2023-09-13T21:03:13.607Z] Azure.Messaging.ServiceBus: The lock supplied is invalid. Either the lock expired, or the message has already been removed from the queue. For more information please see https://aka.ms/ServiceBusExceptions . Reference:ced3aedb-c4da-4f07-9924-c49f4b566262, TrackingId:c87a4b560000048f082601b265022390_G8_B25, SystemTracker:G8:444878248:amqps://mmcsharedservicebus.servicebus.windows.n
et/-390fda99;0:5:6:source(address:/testqueue,filter:[]), Timestamp:2023-09-13T21:04:13 (MessageLockLost). For troubleshooting information, see https://aka.ms/azsdk/net/servicebus/exceptions/troubleshoot.
[2023-09-13T21:03:43.471Z] C# waited 30sec processed message: 1
[2023-09-13T21:03:43.476Z] Executed 'ServiceBusQueueTrigger1' (Succeeded, Id=58f691b8-d49d-49ec-887a-22eb940eb7ba, Duration=90377ms)
[2023-09-13T21:03:43.591Z] C# ServiceBus queue trigger function processed message: 3
[2023-09-13T21:03:43.779Z] Message processing error (Action=Complete, EntityPath=testqueue, Endpoint=MMCSharedServiceBus.servicebus.windows.net)
[2023-09-13T21:03:43.785Z] Azure.Messaging.ServiceBus: The lock supplied is invalid. Either the lock expired, or the message has already been removed from the queue. For more information please see https://aka.ms/ServiceBusExceptions . Reference:ce4299f7-3876-4074-8cf6-39f5f1a56e93, TrackingId:c87a4b560000048f082601b265022390_G8_B25, SystemTracker:G8:444878248:amqps://mmcsharedservicebus.servicebus.windows.n
et/-390fda99;0:5:6:source(address:/testqueue,filter:[]), Timestamp:2023-09-13T21:04:43 (MessageLockLost). For troubleshooting information, see https://aka.ms/azsdk/net/servicebus/exceptions/troubleshoot.

```

**Comment by BorgPrincess** [comment]
Wow.
A year since the last new contribution to the discussion, and more than two years since anybody from Microsoft answered.

Dear MS devs (@paulbatum, @fabiocav, @mathewc, @NickCraver, @kaibocai, @heppersonmicrosoft, @pragnagopa, @siddharth-ms, @liliankasem, @alrod, @balag0 among others) - have another look at this - also for non-.Net FnApps?
Pretty please? 🥺

**Comment by Coops9** [comment]
bump

**Comment by MichaMican** [comment]
Funnily enough the `singleton` settings object in the `host.json` doesn't explain anything about the properties just links to this (still open) issue lol.

https://learn.microsoft.com/en-us/azure/azure-functions/functions-host-json#singleton

EDIT: Just realised that this "problem" was mentioned in a comment before sourceing from even a different MS documentation: https://github.com/Azure/azure-functions-host/issues/912#issuecomment-1499243399

**Comment by mkeinemans** [comment]
bump

**Comment by timtucker-dte** [comment]
The needs here generalize to a lot more than just singleton support -- being able to generically set a limit PER FUNCTION for the max # of parallel executions would be great.

Example:

3 service bus triggers

1. Trigger A should be a singleton
2. Trigger B should run no more than 4 executions in parallel
3. Trigger C is OK to scale up to as many instances as the app can scale to handle

Right now I haven't found a great way of doing all 3 except for deploying separate function apps.

**Comment by jaltin** [comment]
@timtucker-dte I made a more detailed explanation and implementation suggestion earlier in this thread, and I think that your use cases should be covered if that would be implemented.

[https://github.com/Azure/azure-functions-host/issues/912#issuecomment-1059120967](url)

I a still hoping that one day MS will pick up on this!

**Comment by marcselman** [comment]
Yes, this feature could immediately solve an issue we have.

**Comment by mkeinemans** [comment]
> The needs here generalize to a lot more than just singleton support -- being able to generically set a limit PER FUNCTION for the max # of parallel executions would be great.
> 
> Example:
> 
> 3 service bus triggers
> 
>     1. Trigger A should be a singleton
> 
>     2. Trigger B should run no more than 4 executions in parallel
> 
>     3. Trigger C is OK to scale up to as many instances as the app can scale to handle
> 
> 
> Right now I haven't found a great way of doing all 3 except for deploying separate function apps.

@timtucker-dte did you find a way to get the 1st example working?

**Comment by timtucker-dte** [comment]
> > The needs here generalize to a lot more than just singleton support -- being able to generically set a limit PER FUNCTION for the max # of parallel executions would be great.
> 
> [@timtucker-dte](https://github.com/timtucker-dte) did you find a way to get the 1st example working?

Most all our use cases for singletons are handled pretty well by timers, so we haven't focused too much on that.

We struggle a lot more with rate limiting.

A more typical scenario for us:
- Cosmos DB with 10,000 RU/s provisioned
- HTTP functions that should be served as fast as possible and never fail due to 429 errors from running out of RUs
- Service bus functions that should be processed as fast as possible, but should use no more than 8,000 RU/s total
- Service bus functions that need to write data to the DB "eventually" that should use no more than 8,000 RU/s total and run as fast as possible as long as they don't cause either of the scenarios above to fail

Being able to control the # of simultaneous executions on a per-function basis would help a lot, but is still a pretty brute-force solutions to the underlying problem.

At least some of the issues above would be addressed if Cosmos had a "wait to write this until there are free RUs" option in the SDK, but queue-based leveling is still the current pattern recommended by MS:
https://learn.microsoft.com/en-us/azure/architecture/patterns/queue-based-load-leveling

**Comment by timtucker-dte** [comment]
Realizing that this becomes even more important with the current limitations of the Flex Consumption plan -- with the inability to set the max # of instances below 40, setting max instances isn't a viable way to limit parallel execution.

**Comment by BorgPrincess** [comment]
Another 9 months without so much as a sign of life from MS. 

Dear MS devs (@paulbatum, @fabiocav, @mathewc, @NickCraver, @kaibocai, @heppersonmicrosoft, @pragnagopa, @siddharth-ms, @liliankasem, @alrod, @balag0 among others), could _anybody_ please comment?!  🥺 

Does anybody else have _any_ idea how to limit a service bus triggered Python function app on a Flex Consumption plan to no more than one instance running, and no more than one message being processed at a time?

Please?!

**Comment by drdamour** [comment]
> Does anybody else have _any_ idea how to limit a service bus triggered Python function app on a Flex Consumption plan to no more than one instance running, and no more than one message being processed at a time?
> 
> Please?!

In this thread https://github.com/Azure/azure-functions-host/issues/912#issuecomment-1646601602

**Comment by AartBluestoke** [comment]
i've had to write a 'singleton' annotation of my own. - lock/release the lock on an azure blob - and set a background thread to renew all the singleton locks on a slow timer.
As a nice side effect it can take a lock on any value, so multiple functions can share a single lock (ie, storage account and a http trigger that accepts a filename can share a singleton lock) - and the annotation can also choose to reject the trigger or sleep/wait for the lock to free...

i highly recommend your favourite AI to write the boilerplate annotation for the core of what you need
